Mathlib/LinearAlgebra/TensorProduct/Graded/External.lean

```diff
@@ -0,0 +1,270 @@
+  refine (TensorProduct.assoc R _ _ _).toLinearMap
+    ∘ₗ ?_ ∘ₗ (TensorProduct.assoc R _ _ _).symm.toLinearMap
+  refine TensorProduct.map ?_ LinearMap.id
+  exact (gradedComm _ _ _).toLinearMap
+
+theorem tmul_of_gradedMul_of_tmul (j₁ i₂ : ι)
+    (a₁ : ⨁ i, 𝒜 i) (b₁ : ℬ j₁) (a₂ : 𝒜 i₂) (b₂ : ⨁ i, ℬ i) :
+    gradedMul R 𝒜 ℬ (a₁ ⊗ₜ lof R _ ℬ j₁ b₁) (lof R _ 𝒜 i₂ a₂ ⊗ₜ b₂) =
+      (-1 : ℤˣ)^(j₁ * i₂) • ((a₁ * lof R _ 𝒜 _ a₂) ⊗ₜ (lof R _ ℬ _ b₁ * b₂)) := by
+  rw [gradedMul]
+  dsimp only [curry_apply, LinearMap.coe_comp, LinearEquiv.coe_coe, Function.comp_apply, assoc_tmul,
+    map_tmul, LinearMap.id_coe, id_eq, assoc_symm_tmul]
+  rw [mul_comm j₁ i₂, gradedComm_of_tmul_of]
+  -- the tower smul lemmas elaborate too slowly
+  rw [Units.smul_def, Units.smul_def, zsmul_eq_smul_cast R, zsmul_eq_smul_cast R]
+  rw [← smul_tmul', map_smul, tmul_smul, map_smul, map_smul]
+  dsimp
+
+variable {R}
+
+theorem algebraMap_gradedMul (r : R) (x : (⨁ i, 𝒜 i) ⊗[R] (⨁ i, ℬ i)) :
+    gradedMul R 𝒜 ℬ (algebraMap R _ r ⊗ₜ 1) x = r • x := by
+  suffices gradedMul R 𝒜 ℬ (algebraMap R _ r ⊗ₜ 1) = DistribMulAction.toLinearMap R _ r by
+    exact FunLike.congr_fun this x
+  ext ia a ib b
+  dsimp
+  erw [tmul_of_gradedMul_of_tmul]
+  rw [zero_mul, uzpow_zero, one_smul, smul_tmul']
+  erw [one_mul, _root_.Algebra.smul_def]
+
+theorem one_gradedMul (x : (⨁ i, 𝒜 i) ⊗[R] (⨁ i, ℬ i)) :
+    gradedMul R 𝒜 ℬ 1 x = x := by
+  simpa only [_root_.map_one, one_smul] using algebraMap_gradedMul 𝒜 ℬ 1 x
+
+theorem gradedMul_algebraMap (x : (⨁ i, 𝒜 i) ⊗[R] (⨁ i, ℬ i)) (r : R) :
+    gradedMul R 𝒜 ℬ x (algebraMap R _ r ⊗ₜ 1) = r • x := by
+  suffices (gradedMul R 𝒜 ℬ).flip (algebraMap R _ r ⊗ₜ 1) = DistribMulAction.toLinearMap R _ r by
+    exact FunLike.congr_fun this x
+  ext
+  dsimp
+  erw [tmul_of_gradedMul_of_tmul]
+  rw [mul_zero, uzpow_zero, one_smul, smul_tmul']
+  erw [mul_one, _root_.Algebra.smul_def, Algebra.commutes]
+  rfl
+
+theorem gradedMul_one (x : (⨁ i, 𝒜 i) ⊗[R] (⨁ i, ℬ i)) :
+    gradedMul R 𝒜 ℬ x 1 = x := by
+  simpa only [_root_.map_one, one_smul] using gradedMul_algebraMap 𝒜 ℬ x 1
+
+theorem gradedMul_assoc (x y z : DirectSum _ 𝒜 ⊗[R] DirectSum _ ℬ) :
+    gradedMul R 𝒜 ℬ (gradedMul R 𝒜 ℬ x y) z = gradedMul R 𝒜 ℬ x (gradedMul R 𝒜 ℬ y z) := by
+  let mA := gradedMul R 𝒜 ℬ
+    -- restate as an equality of morphisms so that we can use `ext`
+  suffices LinearMap.llcomp R _ _ _ mA ∘ₗ mA =
+      (LinearMap.llcomp R _ _ _ LinearMap.lflip <| LinearMap.llcomp R _ _ _ mA.flip ∘ₗ mA).flip by
+    exact FunLike.congr_fun (FunLike.congr_fun (FunLike.congr_fun this x) y) z
+  ext ixa xa ixb xb iya ya iyb yb iza za izb zb
+  dsimp
+  simp_rw [tmul_of_gradedMul_of_tmul, Units.smul_def, zsmul_eq_smul_cast R,
+    LinearMap.map_smul₂, LinearMap.map_smul, DirectSum.lof_eq_of, DirectSum.of_mul_of,
+    ← DirectSum.lof_eq_of R, tmul_of_gradedMul_of_tmul, DirectSum.lof_eq_of, ← DirectSum.of_mul_of,
+    ← DirectSum.lof_eq_of R, mul_assoc]
+  save
```

Comment:
I assume this was for development only and that we shouldn't be committing `save` in final code?
```suggestion
```